documentation/doxygen/CMakeLists.txt

```diff
@@ -0,0 +1,397 @@
+  #~ ../../core/foundation/
+  #~ ../../core/gui/
+  #~ ../../core/macosx/
+  #~ ../../core/meta/
+  #~ ../../core/metacling/
+  #~ ../../core/clingutils/
+  #~ ../../core/multiproc/
+  #~ ../../core/rint/
+  #~ ../../core/testsupport/
+  #~ ../../core/thread/
+  #~ ../../core/unix/
+  #~ ../../core/winnt/
+  #~ ../../core/imt/
+  #~ ../../core/zip/inc/Compression.h
+  #~ ../../geom/
+  #~ ../../graf2d/asimage/
+  #~ ../../graf2d/cocoa/
+  #~ ../../graf2d/fitsio/
+  #~ ../../graf2d/gpad/
+  #~ ../../graf2d/gpadv7/
+  #~ ../../graf2d/graf/
+  #~ ../../graf2d/gviz/
+  #~ ../../graf2d/postscript/
+  #~ ../../graf2d/quartz/
+  #~ ../../graf2d/win32gdk/
+  #~ ../../graf2d/x11/
+  #~ ../../graf2d/x11ttf/
+  #~ ../../graf3d/eve/
+  #~ ../../graf3d/eve7/
+  #~ ../../graf3d/g3d/
+  #~ ../../graf3d/gl/
+  #~ ../../graf3d/gviz3d/
+  #~ ../../gui/
+  #~ ../../hist/
+  #~ ../../html/
+  #~ ../../io/doc/TFile
+  #~ ../../io/dcache/
+  #~ ../../io/io/
+  #~ ../../io/sql/
+  #~ ../../io/xml/
+  #~ ../../io/xmlparser/
+  #~ ../../main/src/hadd.cxx
+  #~ ../../math/
+  #~ ../../montecarlo/
+  #~ ../../net/auth/
+  #~ ../../net/davix/
+  #~ ../../net/doc/
+  #~ ../../net/http/
+  #~ ../../net/httpsniff/
+  #~ ../../net/net/
+  #~ ../../net/netxng/
+  #~ ../../proof/
+  #~ ../../tmva/
+  #~ ../../roofit/
+  #~ ../../tree/
+  #~ ../../sql/
+  #~ ../../tutorials/
+  #~ ../../bindings/tpython/
+  #~ ../../bindings/pyroot/
+  #~ ../../bindings/r/
+
+  CACHE STRING "Doxyfile input files or folders")
+
+add_custom_target(Preparation
+  COMMAND mkdir -p ${DOCU_LOCATION}/macros ${DOCU_LOCATION}/images ${DOCU_LOCATION}/notebooks
+
+  COMMAND scp ${CMAKE_CURRENT_SOURCE_DIR}/libs.C ./
+  COMMAND scp ${CMAKE_CURRENT_SOURCE_DIR}/makeimage.C ./
+  COMMAND scp ${CMAKE_CURRENT_SOURCE_DIR}/MakeTCanvasJS.C ./
+  COMMAND scp ${CMAKE_CURRENT_SOURCE_DIR}/MakeRCanvasJS.C ./
+  COMMAND scp ${CMAKE_CURRENT_SOURCE_DIR}/makeCollaborationDiagrams.sh ./
+  COMMAND scp ${CMAKE_CURRENT_SOURCE_DIR}/makeNotebooks.sh ./
+  COMMAND scp ${CMAKE_CURRENT_SOURCE_DIR}/modifyClassWebpages.sh ./
+  COMMAND scp ${CMAKE_CURRENT_SOURCE_DIR}/modifyClassWebpage.sh ./
+  COMMAND scp ${CMAKE_CURRENT_SOURCE_DIR}/filter.cxx ./
+
+  COMMAND scp -p -r ${CMAKE_CURRENT_SOURCE_DIR}/images/* ${DOCU_LOCATION}/images/
+  COMMAND tar xfz ${CMAKE_CURRENT_SOURCE_DIR}/mathjax.tar.gz -C ${DOCU_LOCATION}/
+  COMMAND scp -p -r ${CMAKE_CURRENT_SOURCE_DIR}/../../js ${DOCU_LOCATION}/
+
+  COMMAND ${CMAKE_CURRENT_SOURCE_DIR}/makehtmlfooter.sh > ${CMAKE_CURRENT_BINARY_DIR}/htmlfooter.html
+
+  COMMAND ${ROOT_CXX_COMPILER} -DROOT_COMMAND=\\\"${ROOT_root_CMD}\\\" -DDOXYGEN_OUTPUT_DIRECTORY=\\\"${DOCU_LOCATION}\\\" -DDOXYGEN_SOURCE_DIRECTORY=\\\"${CMAKE_CURRENT_SOURCE_DIR}/../../\\\" -DPYTHON_EXECUTABLE=\\\"${PYTHON_EXECUTABLE}\\\" -DPYSPARK_PYTHON=\\\"${PYTHON_EXECUTABLE}\\\" -DLD_LIBRARY_PATH=\\\"${LD_LIBRARY_PATH}:${ROOT_DIR}/lib\\\" -DCMAKE_BUILD_DIRECTORY=\\\"${CMAKE_CURRENT_BINARY_DIR}\\\" -o filter filter.cxx -std=c++17 -O2
```

Comment:
@couet maybe try in line 182 something like this:

```
if(APPLE)
set(EXTRA_INCLUDE_PATH "-I/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.toolchain/usr/....wherever-unistd.h-is-usually-located-inMac")
else()
set(EXTRA_INCLUDE_PATH "")
endif()
```

and then in line 202 end the command with ${EXTRA_INCLUDE_PATH}

I have no macos, so I cannot test it myself, sorry.

Reply:
I do not understand why it was compiling before and not anymore ? This file has not changed a lot (the include was already there) . in the old Makefile we simply did:
```
	`root-config --cxx` -o filter filter.cxx -std=c++14 -O2
```
and it was ok.

Reply:
I am a bit reluctant to do something special for Mac as it was not necessary in the past

Reply:
The difference is that before, you had called thisroot.sh, before building the docu, which adds the Apple include-dirs to the global CXX include paths via environment variables, I believe. Now, we are building the docu as a part of the build step, so before thisroot.sh is available. So we need to adapt the environment variables accordingly.

Reply:
You can also do: make VERBOSE=1
To verify what the failing command is doing exactly. Maybe it's picking a wrong compiler such as clang instead of c++ ?

Reply:
When you will build the documentation (the oxygen command) you will need to have root defined so you will need to call `thisroot.sh` . This is mandatory. So why not calling thisroot.sh before compiling filter.cxx ?

Reply:
You said:

>  Now, we are building the docu as a part of the build step, so before thisroot.sh is available

But when I ran `make -j8` I see that 1st you build root and then you build the doc... so once root is build `thisroot.h`
is available, therefore you can `source` it before compiling filter.cxx and executing `doxygen`.  I do not see why that's would not be possible.

Reply:
But CMake has two parts: Configure and Make. The Configure part is executed at the very beginning, and that's when the relevant commands are defined, so we cannot rely on later at Build-Step to 'overwrite' the Configure-Step commands. So what we do instead is just rely on the existence of root.exe for the Build-Step, but not on calling thisroot.sh
This also makes the environment cleaner. It's working well on Linux. Maybe one or two minor tweaks are needed for Mac.

Reply:
I know, I am speaking of the Make part. In that [art you first build root and then run doxygen (I saw that). So before running doxygen root is ready. the steps done by make are:

1) build root
2) build filter
3) run doxygen

for step 3) you will need to have thisroot.sh done ... **That's mandatory** because you what to build the ref guide for the ROOT just built... no choice... so, as root is ready after step 1) why not running thisroot.sh before step 2) as you will need it anyway in step 3)/

Reply:
This PR was designed to supports
 two ways of building the docu:

1) what you say, ie first building ROOT, then calling thisroot.sh, then calling cmake directly on the CMakelists.txt inside documentation/doxygen

2) builtin, ie a part of the ROOT build step. In this case we do not call thisroot.sh It's not mandatory and it still works perfectly on Linux even for step 3.

I think that maybe you prefer alternative 2, whereas I prefer alternative 1. Both are supported. So please try instead alternative 2. Just build usual ROOT without builtin-docu. Then call thisroot, and in a different folder call CMake towards the documentation/doxygen folder. That will match better with the old Make behavior.

Reply:
I checked the thisroot.sh script and I don't see how it could be affecting the CPLUS_INCLUDE_PATH variable, so not sure whether this would help at all for solving the seen compilation error. Maybe the issue is rather that CMAKE_CXX_COMPILER does not match with Xcode?